pkg/cluster/internal/providers/common/util.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+// AddressPortProtocolMap is used to keep track of declared listen address,
+// port and protocol in node extra port mappings.
+type AddressPortProtocolMap map[string]map[int32]map[string]struct{}
```

Comment:
Do you mean like defining AddressPortProtocolMap structure with the nested map as a property and calling  pointer receivers on AddressPortProtocolMap ?

Reply:
never mind.

Reply:
I don't think this type is necessary at all. You can do a sets.String of protocol + net.JoinHostPort(hostAddr, hostPort). It will take less memory and less code.

Reply:
https://github.com/kubernetes-sigs/kind/tree/main/pkg/internal/sets

Reply:
we should probably round trip through parsing the IPs before joining them because there's multiple ways to represent the same IP address

Reply:
should I use [ParseIPSloppy](https://github.com/kubernetes/utils/blob/d93618cff8a2/net/parse.go#L27) or [ParseIP](https://pkg.go.dev/net#ParseIP) is good enough for this ?